src/builder.rs

```diff
@@ -1,144 +1,275 @@
+        self.output_owners.insert(output.public_key().into(), owner);
+        self.material.outputs.push(output);
+        self
+    }
+
+    /// add a list of outputs
+    pub fn add_outputs(mut self, outputs: impl IntoIterator<Item = (Output, OwnerOnce)>) -> Self {
+        for (output, owner) in outputs.into_iter() {
+            self = self.add_output(output, owner);
+        }
+        self
+    }
+
+    /// get a list of input owners
+    pub fn input_owners(&self) -> Vec<PublicKeyBlst> {
+        self.material.public_keys()
+    }
+
+    /// get sum of input amounts
+    pub fn inputs_amount_sum(&self) -> Amount {
+        self.material
+            .inputs
             .iter()
-            .enumerate()
-            .map(|(out_idx, output)| {
-                let blinding_factor = DbcContent::calc_blinding_factor(
-                    out_idx == self.outputs.len() - 1,
-                    inputs_bf_sum,
-                    outputs_bf_sum,
-                );
-                outputs_bf_sum += blinding_factor;
+            .map(|m| m.true_input.revealed_commitment.value)
+            .sum()
+    }
 
-                let dbc_content = DbcContent::new(
-                    parents.clone(),
-                    output.amount,
-                    output.owner,
-                    blinding_factor,
-                )?;
-                Ok((dbc_content, output.owner))
-            })
-            .collect::<Result<Vec<_>>>()?;
+    /// get sum of output amounts
+    pub fn outputs_amount_sum(&self) -> Amount {
+        self.material.outputs.iter().map(|o| o.amount).sum()
+    }
 
-        let inputs = HashSet::from_iter(self.inputs.into_keys());
-        let outputs = HashSet::from_iter(outputs_and_owners.into_iter().map(|(o, _)| o));
-        Ok(ReissueTransaction { inputs, outputs })
+    /// build a RingCtTransaction and associated secrets
+    pub fn build(
+        self,
+        rng: impl RngCore + rand_core::CryptoRng,
+    ) -> Result<(
+        RingCtTransaction,
+        Vec<RevealedCommitment>,
+        RingCtMaterial,
+        OutputOwnerMap,
+    )> {
+        let result: Result<(RingCtTransaction, Vec<RevealedCommitment>)> =
+            self.material.sign(rng).map_err(|e| e.into());
+        let (transaction, revealed_commitments) = result?;
+        Ok((
+            transaction,
+            revealed_commitments,
+            self.material,
+            self.output_owners,
+        ))
     }
 }
 
-/// Builds a ReissueRequest from a ReissueTransaction and
+/// Builds a ReissueRequest from a RingCtTransaction and
 /// any number of (input) DBC spent proof shares.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Debug)]
 pub struct ReissueRequestBuilder {
-    pub reissue_transaction: ReissueTransaction,
-    pub spent_proof_shares: BTreeMap<SpendKey, HashSet<SpentProofShare>>,
+    pub transaction: RingCtTransaction,
+    pub spent_proof_shares: BTreeMap<usize, HashSet<SpentProofShare>>,
```

Comment:
This is done because the commitments must be in original input order for this code in TransactionValidator to work:
```
        let public_commitments: Vec<Vec<Commitment>> = spent_proofs_sorted
            .iter()
            .map(|s| s.public_commitments.clone())
            .collect();

        transaction.verify(&public_commitments)?;
```
Looking at it again, I should be able to get them in proper order by matching with the tx.mlsags inputs, which are ordered.   anyway, it's a big enough change I will do it in a separate commit, 2nd pass.

Reply:
done.  in 08e5883755c5707d4be626f494bcd60d9e89abfa